Compile only changed exercises.

jeeger opened the pull request:
Check the git commit for all exercises that have been changed, and
skip compiling if the current exercise is not among them.

Fixes #746 .

jeeger commented:
Now how do I get travis to pull the CI scripts from this branch instead of master?

jeeger commented on bin/test-all-examples:
```diff
@@ -9,6 +9,22 @@ fi
 
 exercisedir=$1
 
+if [ "$current_branch" != "master" ]; then
```
I did not modify `test-examples`, only `test-all-examples`. The reasoning being if an exercise is not modified, we want to skip all examples. Thus, testing individually in `test-examples` can be "lifted" into `test-all-examples`.

I hope I didn't misunderstand the purpose of the examples here, though.

petertseng replied:
While I did somehow misunderstand and thought that `test-example` had been used, this now leaves an interesting question. Since `test-example` is not changed, that means if someone were to choose to directly run `test-example` on one individual example, it would not benefit from this addition.

But I suppose that's rare enough that we don't need to think about it. In that case, I have no further change to suggest in this vein.

jeeger replied:
Are these scripts meant to be run manually? In this case, this would require the addition of something like a `--force` switch.

If they're only used by travis, I don't think it's a big problem, as running a `test-example` explicitly indicates strongly that you'd want those examples tested.

jeeger replied:
Ah, but this would require all the travis environment variables. I'll make the script runnable in manual mode as well.

jeeger commented:
Thanks for the information on Travis! I've set it up on my repo now, and I'll do some testing on whether this actually works, or why it doesn't. I'll take this PR off WIP when I'm done.

jeeger commented:
The git commands in the current version don't work correctly, because Travis works on detached head. Now trying to use the Travis environment variables (`TRAVIS_COMMIT_RANGE`, `TRAVIS_PULL_REQUEST` and `TRAVIS_BRANCH`).

sshine commented:
Just so I'm clear, doesn't this PR address #568's last bullet?

jeeger commented:
Hey! Yes, it would, but I'm kinda stuck at the moment. 

As mentioned above, the variables that would be required to do this in a "proper" fashion are sometimes unset, and I don't know why.

petertseng commented:
> doesn't this PR address #568's last bullet?

That's interesting. For the answer to be "yes", that implies that operations that normally affect all exercises would instead be applied in multiple steps. I wonder how that process will look.

sshine commented:
I'm not sure exactly what you mean by multiple steps. I just read:

> Rethink how we're doing our tests and write a more sustainable solution.

and thought that compiling only changed exercises categorized as rethinking the process of reducing time running Travis for PRs. But you're saying that it mainly covers how to compile *all* exercises in a more optimal fashion?

When would it make sense to compile only affected exercises? During PRs, right?

And when would it otherwise make sense to compile all exercises? As a scheduled task?

petertseng commented:
> > Rethink how we're doing our tests and write a more sustainable solution.
> 
> and thought that compiling only changed exercises categorized as rethinking the process of reducing time running Travis for PRs. But you're saying that it mainly covers how to compile all exercises in a more optimal fashion?

Both things you mentioned sound like desirable goals. And not only are they desirable, I predict that we will have need of both of them.

> When would it make sense to compile only affected exercises? During PRs, right?

Yes, that would be great to reduce the amount of time that we must wait to get the Travis CI result back.

Of course, that implies that for PRs that affect all of exercises, this particular measure results in no reduction. We should figure out how, if at all, we want to deal with that.

> And when would it otherwise make sense to compile all exercises? As a scheduled task?

Yes, just as the master branch is run every week, that we can see in https://travis-ci.org/exercism/haskell/builds .

Why is it not instead being run every time a new commit appears on master? Well, they were turned off in https://github.com/exercism/discussions/issues/174#issuecomment-318855073, though we do have the power to turn that back on if so desired.

jeeger commented:
I would be fine with this, as the general structure of the solution is already in this merge request.

Two important points for the discussion, however:
First, I don't think the compilation and testing of the actual exercises uses much time. From me looking at the Travis output, the major time during the build was spent setting up stack and GHC, and building the required dependencies. This would not change with a testing mechanism runs only changed exercises. Using some prebuilt stack LTS images could help here (for more information, see [travis documentation](https://docs.travis-ci.com/user/docker/) and [stack docker features](https://docs.haskellstack.org/en/stable/docker_integration/).

Second:
> > doesn't this PR address #568's last bullet?
> That's interesting. 

The implementation works as follows: When Travis runs, `test-stubs` and `test-all-examples` run for every exercise. These scripts then check whether the current exercise was changed in the commit range this build is testing (theoretically available at `TRAVIS_BRANCH`, `TRAVIS_PULL_REQUEST` and `TRAVIS_COMMIT_RANGE`). If not, they exit immediately. This way, running `test-example` manually will still compile the exercise selected, even if nothing has changed.

jeeger commented:
Alright, I've implemented this and it's working.

One caveat of this solution is that it only marks exercises as changed when a file in the directory is changed. Any changes in the build system that are not in the exercise directory won't cause the exercises to be rebuilt. I've added an exception for the master branch – on the master branch, all exercises are built.

petertseng commented:
I acknowledge that I've been a lacking collaborator in my not providing timely review of this.

In general, the motivation for this PR is greatest whenever we feel the pain from long build times. It appears that we are in a period where we aren't greatly feeling the pain, therefore the motivation to review and merge this gets lower. It's not really anyone's fault, but I'd like to point out the situation and hopefully explain why it shouldn't be discouraging.

I'm very interested in the statement:

> First, I don't think the compilation and testing of the actual exercises uses much time.
> From me looking at the Travis output, the major time during the build was spent setting up stack and GHC, and building the required dependencies.
> This would not change with a testing mechanism runs only changed exercises.

If we find this to be true, we might judge that the benefit from this PR isn't worth the increased complexity in the scripts!

> Using some prebuilt stack LTS images could help here

That is worth exploring, when an interested party has time. I have recorded it in #568 so that we don't forget about it.

I see a comment about `language: bash`

> This is a change that's not strictly necessary, but will probably speed up the build by about a minute or so because rubygems is not updated automatically when Travis starts.

Hm. Sounds like something we could use, and it could be extracted into its own PR. Let's explore that avenue immediately.

jeeger commented:
> I'm very interested in the statement:
> …
> If we find this to be true, we might judge that the benefit from this PR isn't worth the increased complexity in the scripts!

I agree. The shell code used to implement change detection is brittle.

> > Using some prebuilt stack LTS images could help here
> 
> That is worth exploring, when an interested party has time. I have recorded it in #568 so that we don't forget about it.

It looks like a rather simple change as well, judging from the stack documentation. But I'm not familiar with the exercism infrastructure.

> I see a comment about `language: bash`
> Hm. Sounds like something we could use, and it could be extracted into its own PR. Let's explore that avenue immediately.

I'll extract this change into it's own pull request.

petertseng commented:
> I'll extract this change into it's own pull request.

Don't worry, I had already done so in #803 , and it got merged, thank you! Since it was a cherry-pick the original Author line (yours) remains on the commit.

jeeger commented:
Alright, I don't think we can get anything else from this PR. @petertseng I guess we can close it?

petertseng commented:
At your request. We'll revisit if circumstances change such that this does provide a benefit worth the cost. As can be seen in #804 I tested it to put numbers on it.

I thank you for exploring this and making the recommendations based on what you saw!
